File: src/wfrp-helper-web/src/model/creature/CreatureDeterminants.js
import Profession                     from "../professions/Profession";
import Determinant, {DeterminantType} from "../rule/Determinant";

export default class CreatureDeterminants {
    determinants: Determinant[] = [];

    constructor() {
        const speed = new Determinant();
        speed.type = DeterminantType.SPEED;
        const battle = new Determinant();
        battle.type = DeterminantType.BATTLE;
        const shooting = new Determinant();
        shooting.type = DeterminantType.SHOOTING;
        const strength = new Determinant();
        strength.type = DeterminantType.STRENGTH;
        const durability = new Determinant();
        durability.type = DeterminantType.DURABILITY;
        const health = new Determinant();
        health.type = DeterminantType.HEALTH;
        const attack = new Determinant();
        attack.type = DeterminantType.ATTACK;
        const initiative = new Determinant();
        initiative.type = DeterminantType.INITIATIVE;
        const dexterity = new Determinant();
        dexterity.type = DeterminantType.DEXTERITY;
        const leaderSkills = new Determinant();
        leaderSkills.type = DeterminantType.LEADER_SKILLS;
        const intelligence = new Determinant();
        intelligence.type = DeterminantType.INTELLIGENCE;
        const control = new Determinant();
        control.type = DeterminantType.CONTROL;
        const will = new Determinant();
        will.type = DeterminantType.WILL;
        const charisma = new Determinant();
        charisma.type = DeterminantType.CHARISMA;
        this.determinants.push(speed, battle, shooting, strength, durability, health, attack, initiative,
            dexterity, leaderSkills, intelligence, control, will, charisma);
    }

    updateWith(determinants: CreatureDeterminants): CreatureDeterminants {
        if (determinants.determinants) {
            determinants.determinants.forEach(determinant => {
                const newDeterminant = new Determinant();
                newDeterminant.updateWith(determinant);
                this.replaceDeterminant(newDeterminant);
            });
        }
        return this;
    }

    updateDeterminantValue(value: number, type: string) {
        this.findDeterminant(type).value = value;
    }

    findDeterminant(type: string) {
        return this.determinants.find(determinant => type === determinant.type)
    }

    replaceDeterminant(determinant: Determinant) {
        const toReplace = this.findDeterminant(determinant.type);
        let determinantIndex = this.determinants.indexOf(toReplace);
        this.determinants.splice(determinantIndex, 1);
        this.determinants.push(determinant);
    }
}

File: src/main/java/pl/khuzzuk/wfrp/helper/model/creature/CreatureDeterminants.java
package pl.khuzzuk.wfrp.helper.model.creature;

import lombok.Getter;
import lombok.Setter;
import org.hibernate.envers.Audited;
import pl.javahello.DTO;
import pl.khuzzuk.wfrp.helper.model.rule.Determinant;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

import static pl.khuzzuk.wfrp.helper.model.rule.DeterminantType.*;

@Getter
@Setter
@Embeddable
@Audited
@DTO
public class CreatureDeterminants {
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinTable(name = "person_determinants",
            schema = "creature",
            joinColumns = @JoinColumn(name = "person_id"),
            inverseJoinColumns = @JoinColumn(name = "determinants_id"))
    private List<Determinant> determinants = new ArrayList<>();

    public static CreatureDeterminants empty() {
        CreatureDeterminants determinants = new CreatureDeterminants();
        determinants.setDeterminants(List.of(
                Determinant.empty(SPEED),
                Determinant.empty(BATTLE),
                Determinant.empty(SHOOTING),
                Determinant.empty(STRENGTH),
                Determinant.empty(DURABILITY),
                Determinant.empty(HEALTH),
                Determinant.empty(INITIATIVE),
                Determinant.empty(ATTACK),
                Determinant.empty(DEXTERITY),
                Determinant.empty(LEADER_SKILLS),
                Determinant.empty(INTELLIGENCE),
                Determinant.empty(CONTROL),
                Determinant.empty(WILL),
                Determinant.empty(CHARISMA)
        ));
        return determinants;
    }
}
